### PR TITLE
line chart: OffscreenCanvas support for Chart

### PR DESCRIPTION
**Pull request by stephanwlee**
OffscreenCanvas conveniently lets us transfer ownership of Canvas DOM
object to a WebWorker thread so we can reduce expensive point operations
to another process to unlock the UI thread. However, when using
OffscreenCanvas, we have to communicate with the main thread using
message asynchronously.

Since it is not possible to use OffscreenCanvas in some browsers, we
need to gracefully fallback to rendering SVG or WebGL on main thread. As
such, we created an abstraction and use `Chart` and let the Angular side
entry point use the right class.

### CODE REVIEW

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart.ts:46]: nit: Maybe `!== RendererType.WEBGL` is more appropriate, in case more
renderer types are added?

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_bridge.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_bridge.ts:101]: No handler here? Seems somewhat important.

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_bridge.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_bridge.ts:68]: I’m confused; `Chart` is an interface, right? How does `new Chart` work?

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_bridge.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_bridge.ts:121]: Please don’t dereference `_`; it’s confusing, since `_` is supposed to
be a black-hole pattern. If you need to dereference it, just give it a
real name: `const axis`?

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart.ts:134]: `break;`, to prevent headaches later?

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_bridge.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_bridge.ts:84]: Heh; cute. But maybe then guard `index + 1 < length` or assert upfront
that `length` is even, to keep `y: undefined` from sneaking in?

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_types.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_types.ts:61]: sp.: *Changed*, not *Changede*

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_types.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_types.ts:43]: Can we call this `UpdateViewBoxMessage`, or something else more
descriptive than just `UpdateMessage`?

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_types.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_types.ts:26]: To make sure I understand: `RESIZE` is for when the DOM extent of the
chart container updates, `UPDATE_VIEW_BOX` is for when the chart window
(in data coordinates) updates?

Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_types.ts:26]: Here you have `FOO_UPDATE`, `BAR_CHANGED`, and `UPDATE_BAZ`. Let’s be
consistent. Any option is fine with me.

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_types.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_types.ts:22]: naming: As the counterpart of *guest*, *host* feels more familiar to me
than *main*. (E.g., virtual machines.) How about `HostToGuestEvent`?

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_utils.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_utils.ts:20]: Why `ArrayBufferLike` rather than `ArrayBuffer`? Doesn’t it have to be
literally an `ArrayBuffer` to be `Transferable`?

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_types.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_types.ts:56]: Why not just pass a `CompactDataSeries` here?

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_utils.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_utils.ts:24]: sp.: *postMessage*, not *psotMessage*

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_utils.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_utils.ts:65]: Using `const points: Array<{x: number, y: number}> = [];` should work,
yes? That way I don’t have to think about the downcast.

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_utils_test.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_utils_test.ts:18]: Nice tests.

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_types.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_types.ts:1]: Maybe a filename like `{message,event}{s,_types}` would be more precise?

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_test.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_test.ts:109]: nit: *compact*, not *compat*

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_test.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_test.ts:52]: Delete?

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_test.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_test.ts:128]: sp.: *Flattened*, not *Flatttened*

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_utils.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_utils.ts:1]: Maybe `compact_data_series` is a better filename here?

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_test.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/worker_chart_test.ts:147]: I don’t object strongly to this, so I’ll leave it up to you, but perhaps
consider using `decompactDataSeries` here rather than poking at the
exact representation? Since the contract is “the host sends you
something that you can decompact to get the right data”, it’d be nice to
test against that specification rather than the current implementation
of the compaction format.

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/compact_data_series.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/compact_data_series.ts:27]: I never actually asked, but: while this seems 100% plausible to me, I’m
curious whether you measured it and, if so, what the figures were
compared to the compaction overhead here?

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/compact_data_series_test.ts
Comment (wchargin) [inline tensorboard/webapp/widgets/line_chart_v2/lib/worker/compact_data_series_test.ts:111]: Optional: might be nice for one test to include fractional numbers that
don’t get just rounded down to zero? You do cover NaN/Infinities, which
seems pretty reasonable, hence optional.

**Review by wchargin — APPROVED** [review]
Approved modulo naming consistency comment.

File: tensorboard/webapp/widgets/line_chart_v2/lib/worker/compact_data_series.ts
Comment (stephanwlee) [reply tensorboard/webapp/widgets/line_chart_v2/lib/worker/compact_data_series.ts:27]: Yup, it was measured. When sending points of 1000 points (`{x: number; y: number}`), I saw about 4.5x. The testing was not extensive and does not include larger data points.

**Pull request merged**
